Mimic Ruby's backtrace logic in ruby interpreter and tracer, add support for ruby CMEs

dalehamel opened the pull request:
# What

This overhauls the ruby interpreter and unwinder to more closely mimic what ruby's own backtrace function does. In particular, it adds support for collecting and symbolizing Callable Method Entries (CMEs), which contain additional information.

This adds support for new features in symbolizing ruby frames:

- Adds class information when possible, in ruby 3.3.0+, fixing https://github.com/open-telemetry/opentelemetry-ebpf-profiler/issues/714

It also fixes an existing bug in ruby https://github.com/open-telemetry/opentelemetry-ebpf-profiler/issues/770

In addition to this, it changes how frames are symbolized:

- For c func method frames, rather than skip the frame we attempt to symbolize it (supported for all versions of ruby)
  - Presuming we will return to ruby symbolization (not true in ruby < 2.6.0), we save the cfunc and push the frame only after returning to ruby. This way the cfunc will "own" the native frames it calls as we interleave the stack
- For iseq-backed method entry frames, we read the label from the correct location and do our best to add classpath information to it
- For plain iseq-backed frames (which is all that was previously supported), we choose between base_label and label more appropriately

## Real world example

This makes the bpf ruby interpreter closer to being usable and helpful with real production workloads.

These examples are taken from a production job server using our internal job system called "Hedwig".

Before:

<img width="912" height="638" alt="Screenshot 2025-10-31 at 11 54 38 AM" src="https://github.com/user-attachments/assets/2383eb67-ecdc-46ed-a7b2-25a84df024f4" />

After:


<img width="912" height="638" alt="Screenshot 2025-10-31 at 11 54 16 AM" src="https://github.com/user-attachments/assets/b1f32fb4-8004-4186-92d7-767bb06742cd" />

(Note: that these are aggregated by a custom backend, and the [rb]:: prefixes are added by that.)

Notice that the names are much more descriptive, including the class name or extended label information that was previously omitted.

Notice also how `Kernel#fork` owns the native frames calling `rb_f_fork`, and likewise for `Array#each` owning `rb_ary_each`. Previously these frames were elided.

These frames much more closely resemble what Ruby developers will be familiar with when looking at stackprof of vernier profiles, as well as backtracie. I recently submitted the same thing for [rbspy](https://github.com/rbspy/rbspy/pull/400) following basically the same approach and algorithm but done in one process, in userspace. In fact, when placed side by side with profiles from stackprof, developers can't tell the difference between the new native profiles:

<img width="1223" height="808" alt="Screenshot 2025-10-31 at 3 19 51 PM" src="https://github.com/user-attachments/assets/2e01e577-ef7e-4aff-9229-87cd572cc154" />

vs stackprof request profiles they are already familiar with:

<img width="928" height="575" alt="Screenshot 2025-10-31 at 3 19 08 PM" src="https://github.com/user-attachments/assets/37431e39-baee-4310-bbcf-8f07353c4b5d" />

But, it is also possible to get more context about what it is doing, since we can see kernel and userspace native frames in the context of the ruby:

<img width="1558" height="1064" alt="image" src="https://github.com/user-attachments/assets/24c0864a-5a6b-48b6-9f96-dcb4934a6020" />



# Why

The current unwinding approach is not consistent with what ruby's own backtrace function does. This is confusing to users who are familiar with ruby profilers like stackprof and vernier, which uses's ruby's built-in profiling stack unwinding.

We also are currently missing information that makes the ruby unwinder useful for any real app in production. Knowing the class information makes it easier to see what is actually going on in the application.

The existing symbolization and recording of stack frames loses a lot of data, it doesn't handle callable method entries at all and doesn't navigate to the same instruction sequences and label values, so it can provide surprising results.

# How

As much as possible, the code mimics the behaviour of ruby's own `rb_profile_frames` from ruby's `vm_backtrace.c`, both for collecting the frames:

- Every frame is captured in the same way. The bulk of this is done by mimic'ing ruby's own `check_method_entry` to walk from the ep to the correct frame. We have to unroll and optimize this for it to work, since BPF doesn't allow a "while" loop.

And when symbolizing:
- When symbolizing the frames, we mimic `rb_profile_frame_full_label` as much as possible, considering the class name to get the qualified method name, and using the label, base_label, and method_name labels to compose the most descriptive possible label regardless of it is a CME or bare ISEQ.
- For c functions, we read the label from the global ID table, imitating ruby's `id2str`

I also added the support for symbolizing frames that are backed by a ruby CME, this adds new code for `id2sym` which is used to look up C method identifiers from the global symbol table.

For both types of CMEs, I've added the code to read the class data from `ext.classpath` on ruby 3.3.0+.

All frames also use the equivalent of `rb_profile_frame_full_name`, regardless of if the class information is available, as this can still provide a more contextful label.


# Reviewer Notes

I'm aware this is a large changeset, especially after we just landed #170. I considered not including the class path and C symbolizing stuff, but decided it would be a bit silly to submit them separately since there isn't much of a point in collecting CMEs if we don't also use them. Because I'm both updating the BPF side (collection, building up frame buffer in the same way rb_profile_frames does by getting CMES), and Go side (symbolizing the results), there is a fair bit to change here.

I recommend using github's "side by side" view to review this, the intermixed one won't make much sense otherwise.

I'd recommend starting it in the BPF code

- read_ruby_frame https://github.com/dalehamel/opentelemetry-ebpf-profiler/pull/5/files#diff-b61481ee49ef1ed454c5d7a7a13853d92f65f6fd56d2d38e475b45ca97a9ab9eR102
- comparing this to the corresponding ruby code in vm_backtrace.c (thread_profile_Frames)
- Review the rest of the C / bpf, most of which is supporting boilerplate stuff like constants for errors and metrics or shared constants with Go.

On then moving to the go side:

- start in Symbolize https://github.com/dalehamel/opentelemetry-ebpf-profiler/pull/5/files#diff-c0a6a39f3d56f7bff3ba5d2a41360932a50b265067b0717cc67916f4ebaf7a62R930
- Compare this with the logic starting in rb_profile_frame_full_name to see why these fields are being collected and to see the algorithm I am trying to copy
- Then review the rest of the go, which is boilerplate stuff and some tests

My apologies for the large diff, I did hold back additional stuff which I want to submit in subsequent PRs:

- Garbage collection state detection
- Ruby JIT support
- Ongoing work to address getting the EC from TLS 

However, this change represents the most important, common base for all the additional work. I tried to keep the changes as small as I could, but adding tests, comments, and boiler plate has pushed the diff larger.

dalehamel commented on interpreter/ruby/ruby.go:
```diff
@@ -628,57 +716,407 @@ func (r *rubyInstance) getRubyLineNo(iseqBody libpf.Address, pc uint64) (uint32,
+		singletonObject := r.rm.Ptr(classAddr + libpf.Address(r.r.vmStructs.rclass_and_rb_classext_t.classext+r.r.vmStructs.rb_classext_struct.as_singleton_class_attached_object))
+		classpathPtr = r.rm.Ptr(singletonObject + libpf.Address(r.r.vmStructs.rclass_and_rb_classext_t.classext+r.r.vmStructs.rb_classext_struct.classpath))
+
+		// TODO (dalehamel) in future PR handle anonymous classes and modules
```
I can remove my handle from this but I basically promise to submit this later, as it is, i feel the PR is big enough and this is kinda an esoteric edge case.

fabled replied:
I think this is ok.

dalehamel commented on tools/coredump/testdata/amd64/ruby-3.5.0-loop.json:
```diff
@@ -4,27 +4,26 @@
-        "<main>+0 in /app/loop.rb:29",
-        "loop+0 in <internal:kernel>:168",
+        "Range#each+0 in <cfunc>:0",
+        "block in <main>+0 in /app/loop.rb:29",
```
Notice that a lot of these now have more context because we have conditional logic picking between "label", "base_label", and "method name"  from the iseq, to match ruby's own logic for this.

dalehamel commented on tools/coredump/testdata/amd64/ruby-3.5.0-loop.json:
```diff
@@ -4,27 +4,26 @@
         "libruby.so.3.5.0+0x31e2ea",
         "libruby.so.3.5.0+0x325240",
         "libruby.so.3.5.0+0x333b86",
-        "<main>+0 in /app/loop.rb:29",
```
This is an example of the repeated frame bug (#770 ) which is now fixed, hence why it doesn't repeat anymore.

ivoanjo commented on tools/coredump/testdata/amd64/ruby-3.3.9-loop.json:
```diff
@@ -4,27 +4,26 @@
-        "is_prime+0 in /app/loop.rb:10",
-        "sum_of_primes+0 in /app/loop.rb:20",
-        "<main>+0 in /app/loop.rb:30",
+        "Object#is_prime+0 in /app/loop.rb:45481984",
```
Minor: I wonder if it's worth setting the line number to 0 or -1 already in this PR, rather than having the incorrect line until the bigger fix is ready.

ivoanjo commented on support/ebpf/ruby_tracer.ebpf.c:
```diff
@@ -17,18 +18,280 @@ struct ruby_procs_t {
+  const void *ep;         // cfp[4] / block[1]
+  const void *block_code; // cfp[5] / block[2] -- iseq, ifunc, or forwarded block handler
+  void *jit_return;       // cfp[6] -- return address for JIT code
+} rb_control_frame_t;
```
Minor: I wonder if it's worth introducing a typedef for `VALUE`, to keep the closer alignment with the types used on the Ruby VM (rather than replacing them with a `void *`)

fabled commented on interpreter/ruby/ruby.go:
```diff
@@ -635,57 +723,407 @@ func (r *rubyInstance) getRubyLineNo(iseqBody libpf.Address, pc uint64) (uint32,
+// qualified_method_name, translated into golang
+// https://github.com/ruby/ruby/blob/v3_4_7/vm_backtrace.c#L1947
+// qualified_method_name, translated into golang
+// https://github.com/ruby/ruby/blob/v3_4_7/vm_backtrace.c#L1947
```
```suggestion
```
duplicate comment

fabled commented on support/ebpf/ruby_tracer.ebpf.c:
```diff
@@ -17,18 +18,280 @@ struct ruby_procs_t {
+#define MAX_EP_CHECKS              6
+
+// Constants related to reading a method entry
+#define VM_METHOD_TYPE_ISEQ 0
```
could we get source references for these?

fabled commented on support/ebpf/ruby_tracer.ebpf.c:
```diff
@@ -83,14 +364,12 @@ static EBPF_INLINE ErrorCode walk_ruby_stack(
           &stack_ptr_current,
           sizeof(stack_ptr_current),
           (void *)(current_ctx_addr + rubyinfo->vm_stack))) {
-      DEBUG_PRINT("ruby: failed to read current stack pointer");
```
Are these removed to just reduce ebpf byte code size?

fabled commented on support/ebpf/ruby_tracer.ebpf.c:
```diff
@@ -17,18 +18,280 @@ struct ruby_procs_t {
+  if (rubyinfo->version < 0x30100)
+    cf_size -= sizeof(control_frame.jit_return);
+
+read_cfp:
```
unused label?

dalehamel replied:
yes it was used an in earlier version. Replacing the label with a comment.

fabled commented on support/ebpf/ruby_tracer.ebpf.c:
```diff
@@ -17,18 +18,280 @@ struct ruby_procs_t {
+// Next iteration of the loop, or error out if we have hit the maximum as we
+// couldn't find the method entry
+next_ep:
+  if (ep_check++ < MAX_EP_CHECKS && (!((u64)vm_env.flags & VM_ENV_FLAG_LOCAL))) {
```
I would really prefer this loop to be a `for` with the proper `UNROLL` annotation. This might end up generating unrolled code at some point (and fail verifier in the old kernels).

In general `goto`s like this suggest this probably should be split into more functions. While there is no rule to not use goto, I personally think they should not be used in complex way like this where it becomes readability issue.

dalehamel commented:
> Thanks for this! And apologies for the delay.

No worries, very happy to see your review as I come back from the winter break.

> Can you also rebase now that the variable length frame PR is merged?

Done, and I've pushed it, but still need to address your comments.

> First glance done, mostly mechanical.

Thanks, i'm working on addressing these now.

> I need to study the Ruby VM also a bit to understand the split what needs to be in ebpf, and what can be done in Go side. 

I've left a lot of documentation breadcrumbs to follow, but basically in BPF we need to unwind the stack, and build up the same "frame buffer" that ruby's own backtrace algorithm does. For "bare iseq" entries the logic is simple, we just push the address of the iseq body. For callable method entries ("cme"), we have to do some additional walking to find that entity. I don't think we can offload any more work into go, I had tried an earlier version that attempted to do the cme walking in go but this proved to be too volatile as the environment pointers and control frame pointers are not stable, but iseq entries and cmes are.

dalehamel commented on interpreter/ruby/ruby.go:
```diff
@@ -635,57 +723,407 @@ func (r *rubyInstance) getRubyLineNo(iseqBody libpf.Address, pc uint64) (uint32,
+	if classMask == rubyTIClass {
+		//https://github.com/ruby/ruby/blob/b627532/vm_backtrace.c#L1931-L1933
+
+		if klassAddr := r.rm.Ptr(classAddr + libpf.Address(r.r.vmStructs.rbasic_struct.klass)); klassAddr != 0 {
```
I think we can reasonably do this for the flags and klass as they are reliably at the start of the struct and at fixed offsets.

For singleton and classpath though they are much further down in the struct (currently ~100 bytes) so unless we read the whole object i'm not sure that makes sense. How expensive is a large read vs multiple smaller reads?

dalehamel commented on interpreter/ruby/ruby.go:
```diff
@@ -635,57 +723,407 @@ func (r *rubyInstance) getRubyLineNo(iseqBody libpf.Address, pc uint64) (uint32,
-	// rb_iseq_constant_body
-	// https://github.com/ruby/ruby/blob/5445e0435260b449decf2ac16f9d09bae3cafe72/vm_core.h#L311
-	iseqBody := libpf.Address(frame.File)
+	lastId := r.rm.Uint32(r.globalSymbolsAddr)
```
Any time a string is interned, this value can change https://github.com/ruby/ruby/blob/20cda200d3ce092571d0b5d342dadca69636cb0f/symbol.c#L795, so yes I'd say that this is potentially volatile.

> Perhaps we could still cache it, and re-read it only if the serial is larger then the cached lastId?

Yes this is probably ok, since it is only being used for the check. We would need to do the check twice but I guess that is cheaper than unconditionally re-reading the same value.

dalehamel replied:
This is now cached

dalehamel commented on support/ebpf/ruby_tracer.ebpf.c:
```diff
@@ -103,120 +382,41 @@ static EBPF_INLINE ErrorCode walk_ruby_stack(
-      DEBUG_PRINT("ruby: failed to push frame");
+    error = read_ruby_frame(record, rubyinfo, stack_ptr, next_unwinder);
+    if (error != ERR_OK)
       return error;
```
I couldn't find an example of other unwinders doing this, but maybe i'm looking for the wrong thing.

Either way, perhaps I could follow up with another PR to make continuing with unwinding more robust once this and the other ruby PRs i have queued up have been reviewed and hopefully merged.

dalehamel commented:
@fabled I believe I've addressed all your first batch of comments with either code or conversation, apologies if I missed anything. I believe this should be ready for another look when you have the time, thanks!

dalehamel commented on support/ebpf/ruby_tracer.ebpf.c:
```diff
@@ -17,18 +18,280 @@ struct ruby_procs_t {
+
+  // Ruby prior to 3.1.0 control frame did not have jit_return
+  if (rubyinfo->version < 0x30100)
+    cf_size -= sizeof(control_frame.jit_return);
```
It turns out we already had this in procinfo 🤦‍♂️ thanks for the tip, i've updated it to just use that value and simplified this code accordingly.

dalehamel commented on interpreter/ruby/ruby.go:
```diff
@@ -635,57 +723,407 @@ func (r *rubyInstance) getRubyLineNo(iseqBody libpf.Address, pc uint64) (uint32,
+		idsLen = uint64((flags & RARRAY_EMBED_LEN_MASK) >> RARRAY_EMBED_LEN_SHIFT)
+	} else {
+		idsPtr = r.rm.Ptr(ids + libpf.Address(vms.rarray_struct.as_heap_ptr))
+		idsLen = r.rm.Uint64(ids + libpf.Address(vms.rarray_struct.as_ary))
```
I added a size of array member to vm structs, and read the whole RArray in one operation.

dalehamel commented on interpreter/ruby/ruby.go:
```diff
@@ -635,57 +723,407 @@ func (r *rubyInstance) getRubyLineNo(iseqBody libpf.Address, pc uint64) (uint32,
+	ids := r.rm.Ptr(r.globalSymbolsAddr + libpf.Address(vms.rb_symbols_t.ids))
+	idx := serial / idEntryUnit
+
+	flags := r.rm.Ptr(ids)
```
It is a ruby VALUE which is technically a uintptr_t https://github.com/ruby/ruby/blob/50b719115af8344888ef2735788ebc9f32c05a54/parser_value.h#L19

Note that elsewhere in the existing code, flags are already read via r.rm.Ptr in a variety of places. Note that in this case of flags, it looks like they just treat it as a uint64. These existing uses must, i guess, not have any `rm.Bias` to subtract.

In any case, i think the correct way to read flags here, and elsewhere, is uint64 as we are not actually returning a pointer, we are just getting the first 8 bytes at that address. I've updated this one accordingly, we'll need to follow-up on the others.

florianl commented on support/ebpf/ruby_tracer.h:
Can we move these into `support/ebpf/ruby_tracer.ebpf.c`?

In `ruby_tracer.ebpf.c` we already have some `#define`s and here are just a small number of `#define`s in this headerfile.

fabled reviewed (commented):
Thanks for the updates! I'd also like to get this in sooner than later. To get things forward, the few things I'd like to request are:
 - let's not do the 48-bit pointer assumption in this PR, I created #1084 to first discuss it, and then generalize it. It's anyway an optimization - not a requirement. If needed, feel free to increase the frame data size instead.
 - make the ruby specific constants have `ruby` somewhere in them to avoid namespace conflicts
 - also few more nits added in the comments

After this, I'm ok to approve this and improve things in follow up PRs which are hopefully smaller. Thanks for your patience and work on this!

dalehamel commented:
> LGTM! One whitespace nit found. 

Fixed and I added a comment about a potential change to the clang format config to catch it in the future.

> You'll need to merge with main due to other merged PRs touching the ebpf code.

Rebased, rebuilt, and pushed.
